### ISSUE TITLE
[Bug] - dnf upgrade fails with kernel-livepatch packages installed

### ISSUE THREAD

**Issue by nosnilmot** [amazonlinux/amazon-linux-2023#745]
**Describe the bug**

With Kernel Live Patching enabled, the default configuration can get in a state where it is impossible to simply `dnf upgrade`.

**To Reproduce**

1. Install Amazon Linux 2023
2. Enable Kernel Live Patching
3. Enable Automatic updates (`systemctl enable --now dnf-automatic-install.timer`)
4. Wait (a while - months)
5. Realize Amazon Linux 2023 does NOT get updates by default
6. Panic
7. Switch to `latest` releasever (`echo latest | sudo tee /etc/dnf/vars/releasever`)
8. Try to update
9. Panic again

**Expected behavior**
`dnf` should install available updates

**Screenshots**
```
$ sudo dnf upgrade
Last metadata expiration check: 9:19:15 ago on Wed Jul  3 20:55:56 2024.
Error: 
 Problem: The operation would result in removing the following protected packages: kernel
(try to add '--skip-broken' to skip uninstallable packages)
```

```
$ sudo dnf -d 10 upgrade
timer: config: 1 ms
Loaded plugins: builddep, changelog, config-manager, copr, debug, debuginfo-install, download, generate_completion_cache, groups-manager, kpatch, needs-restarting, playground, release-notification, repoclosure, repodiff, repograph, repomanage, reposync, supportinfo
DNF version: 4.14.0
Command: dnf -d 10 upgrade 
Installroot: /
Releasever: latest
cachedir: /var/cache/dnf
Base command: upgrade
Extra commands: ['-d', '10', 'upgrade']
User-Agent: constructed: 'libdnf (Amazon Linux 2023; generic; Linux.aarch64)'
repo: using cache for: amazonlinux
amazonlinux: using metadata from Thu Jun 27 00:00:00 2024.
repo: using cache for: kernel-livepatch
kernel-livepatch: using metadata from Mon Jun 24 00:00:00 2024.
Last metadata expiration check: 9:57:13 ago on Wed Jul  3 20:55:56 2024.
timer: sack setup: 145 ms
--> Starting dependency resolution
--> Finished dependency resolution
timer: depsolve: 143 ms

Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/dnf/cli/main.py", line 130, in cli_run
    ret = resolving(cli, base)
  File "/usr/lib/python3.9/site-packages/dnf/cli/main.py", line 166, in resolving
    base.resolve(cli.demands.allow_erasing)
  File "/usr/lib/python3.9/site-packages/dnf/base.py", line 929, in resolve
    raise exc
dnf.exceptions.DepsolveError: 
 Problem: The operation would result in removing the following protected packages: kernel
Error: 
 Problem: The operation would result in removing the following protected packages: kernel
(try to add '--skip-broken' to skip uninstallable packages)
Cleaning up.
```

**Additional context**

It is possible to get *most* updates installed by excluding `kernel` and `kernel-livepatch` packages:

```
$ sudo dnf upgrade --exclude kernel --exclude kernel-livepatch\*
Last metadata expiration check: 9:25:31 ago on Wed Jul  3 20:55:56 2024.
Dependencies resolved.
====

....

Complete!
```

Significantly increasing `installonly_limit` in `/etc/dnf/dnf.conf` (to >= number of installed `kernel-livepatch`, 14, although that may not be the actual constraint) also allows dependency resolution to complete:

```
sudo dnf upgrade
Last metadata expiration check: 10:06:43 ago on Wed Jul  3 20:55:56 2024.
Dependencies resolved.
==========================================================================================
 Package                         Arch     Version                 Repository         Size
==========================================================================================
Installing:
 kernel                          aarch64  6.1.34-56.100.amzn2023  amazonlinux        27 M
 kernel                          aarch64  6.1.34-58.102.amzn2023  amazonlinux        27 M
 kernel                          aarch64  6.1.34-59.116.amzn2023  amazonlinux        27 M
 kernel                          aarch64  6.1.38-59.109.amzn2023  amazonlinux        27 M
 kernel                          aarch64  6.1.49-69.116.amzn2023  amazonlinux        27 M
 kernel                          aarch64  6.1.49-70.116.amzn2023  amazonlinux        27 M
 kernel                          aarch64  6.1.52-71.125.amzn2023  amazonlinux        27 M
Upgrading:
 kernel-livepatch-6.1.29-50.88   aarch64  1.0-2.amzn2023          kernel-livepatch   23 k
 kernel-livepatch-6.1.34-56.100  aarch64  1.0-1.amzn2023          kernel-livepatch   15 k
 kernel-livepatch-6.1.34-58.102  aarch64  1.0-1.amzn2023          kernel-livepatch   15 k
 kernel-livepatch-6.1.34-59.116  aarch64  1.0-1.amzn2023          kernel-livepatch   15 k
 kernel-livepatch-6.1.38-59.109  aarch64  1.0-1.amzn2023          kernel-livepatch   13 k
 kernel-livepatch-6.1.49-69.116  aarch64  1.0-2.amzn2023          kernel-livepatch   51 k
 kernel-livepatch-6.1.49-70.116  aarch64  1.0-2.amzn2023          kernel-livepatch   51 k
 kernel-livepatch-6.1.52-71.125  aarch64  1.0-2.amzn2023          kernel-livepatch   51 k
 kernel-livepatch-6.1.55-75.123  aarch64  1.0-1.amzn2023          kernel-livepatch   27 k

Transaction Summary
==========================================================================================
Install  7 Packages
Upgrade  9 Packages

Total download size: 189 M
Is this ok [y/N]:
```

**Comment by zcobol** [comment]
There's a discussion about AL2023 automatic update strategies at https://github.com/amazonlinux/amazon-linux-2023/discussions/395

**Comment by andreasstieger** [comment]
Getting someone to look at this for triage/action.

**Comment by bjoernd** [comment]
Sorry for the long wait.

> Realize Amazon Linux 2023 does NOT get updates by default

Note, this is intentional as per https://docs.aws.amazon.com/linux/al2023/ug/deterministic-upgrades.html. The idea here is that you should never be "surprised" by a package change that you did not intentionally make.

You can emulate the alternative behavior (always update to latest) with `echo latest | sudo tee /etc/dnf/vars/releasever`.

**Comment by bjoernd** [comment]
I tried reproducing your issue on an x86 AMI and didn't succeed. Trying aarch64 (as per your logs) next.

**Comment by bjoernd** [comment]
I successfully executed these commands starting from `kernel-6.1.94-99.176.amzn2023.aarch64` and ending with `kernel-6.1.102-108.177.amzn2023.aarch64`:

```
sudo dnf upgrade --releasever=latest
sudo dnf install -y kpatch-dnf
sudo dnf kernel-livepatch -y auto
sudo systemctl enable kpatch.service && sudo systemctl start kpatch.service
sudo dnf upgrade --releasever=latest
```

Can you retry if your problem still exists, please?

**Comment by nosnilmot** [comment]
> Note, this is intentional as per https://docs.aws.amazon.com/linux/al2023/ug/deterministic-upgrades.html. The idea here is that you should never be "surprised" by a package change that you did not intentionally make.

I recognise this _now_, but I think it is incredibly poorly documented, as evidenced also by the existence of #395 (and 1+ year on nobody has actually pointed out the correct way to enable automatic updates in that ticket!). In 2024 I didn't think it was conceivable that a modern Linux distribution would be shipping without at least enabling security updates by default.

The description of major and minor releases, where minor releases '... include feature and security updates, but don't include package changes.' certainly implies that updates are available within a release, and I assumed that is what you get out of the box. Actually what it means is no release ever gets any updates, to get updates you have to switch to a newer release.

(anyway, this is a distraction, and I had been intending to avoid this topic on this issue, which is about a specific situation where updates are blocked even when enabled)

> I successfully executed these commands

Sure, updating immediately with the latest updates available at that time is a much simpler situation with no accumulation of multiple updates. I note you also did not enable always updating to latest (only using CLI `--releasever=latest` option), or use `dnf-automatic-install`

> Can you retry if your problem still exists, please?

The instance I originally reported this for did get updated after I accidentally left `installonly_limit` set to 14 and automatic updates kicked in, so it is no longer in the same state.
If my assumption above is correct, reproducing from scratch might take some time - how often are `kernel` & `kernel-livepatch` packages released?

I believe the problem is this:
1. Multiple kernel packages had been installed, each resulting in `kernel-livepatch` packages also getting installed
2. `installonly_limit` causes older unused `kernel` packages to be removed, but the corresponding `kernel-livepatch` packages were NOT uninstalled
3. On update, `dnf` tries to pull in the now-missing `kernel` packages to satisfy requirements of the already-installed `kernel-livepatch` packages, but this causes problems because it also wants to remove those `kernel` packages to satisfy the `installonly_limit` configuration

Right now the instance is stuck in a different daily flip-flop upgrade/downgrade cycle (this has been going on since I opened this ticket). This may be another manifestation of the same problem?

```
================================================================================
 Package                  Arch    Version                Repository        Size
================================================================================
Installing:
 kernel                   aarch64 6.1.55-75.123.amzn2023 amazonlinux       27 M
Upgrading:
 kernel-livepatch-6.1.55-75.123
                          aarch64 1.0-1.amzn2023         kernel-livepatch  27 k
Removing:
 kernel                   aarch64 6.1.29-50.88.amzn2023  @amazonlinux     140 M
Downgrading:
 kernel-livepatch-6.1.29-50.88
                          aarch64 1.0-0.amzn2023         amazonlinux       17 k

Transaction Summary
================================================================================
Install    1 Package
Upgrade    1 Package
Remove     1 Package
Downgrade  1 Package

Updates completed at Mon Aug  5 06:00:38 2024
```
and
```
================================================================================
 Package                  Arch    Version                Repository        Size
================================================================================
Installing:
 kernel                   aarch64 6.1.29-50.88.amzn2023  amazonlinux       26 M
Upgrading:
 kernel-livepatch-6.1.29-50.88
                          aarch64 1.0-2.amzn2023         kernel-livepatch  23 k
Removing:
 kernel                   aarch64 6.1.55-75.123.amzn2023 @amazonlinux     145 M
Downgrading:
 kernel-livepatch-6.1.55-75.123
                          aarch64 1.0-0.amzn2023         amazonlinux       21 k

Transaction Summary
================================================================================
Install    1 Package
Upgrade    1 Package
Remove     1 Package
Downgrade  1 Package

Updates completed at Mon Aug  5 06:01:57 2024
```

if it helps:
```
$ rpm -qa kernel\* | sort
kernel-6.1.102-108.177.amzn2023.aarch64
kernel-6.1.29-50.88.amzn2023.aarch64
kernel-6.1.94-99.176.amzn2023.aarch64
kernel-livepatch-6.1.102-108.177-1.0-0.amzn2023.aarch64
kernel-livepatch-6.1.25-37.47-1.0-0.amzn2023.aarch64
kernel-livepatch-6.1.27-43.48-1.0-0.amzn2023.aarch64
kernel-livepatch-6.1.29-47.49-1.0-0.amzn2023.aarch64
kernel-livepatch-6.1.29-50.88-1.0-2.amzn2023.aarch64
kernel-livepatch-6.1.41-63.114-1.0-0.amzn2023.aarch64
kernel-livepatch-6.1.55-75.123-1.0-0.amzn2023.aarch64
kernel-livepatch-6.1.94-99.176-1.0-0.amzn2023.aarch64
kernel-livepatch-repo-s3-2023.5.20240805-0.amzn2023.noarch
kernel-srpm-macros-1.0-14.amzn2023.0.2.noarch
kernel-tools-6.1.102-108.177.amzn2023.aarch64
```

**Comment by ozbenh** [comment]
This is a feature of Amazon Linux 2023. Deterministic updates is actually extremely useful in the typical cloud use case of large fleets where you want to be able to reason precisely about what versions of pacakges are deployed and be in control of deployment of updates.

This guarantees that if you launch AMI X accross your fleet, and later add new instances with the same AMI they will see exactly the same package set and any package installed on top of said AMI will be the same versions accross the board.

Every single time we release updated packages (including security updates) we essentially create a new "version" of the repositories. That is basically a release. So yes, you need to update to a new "release" to get those (however you don't necessarily have to ingest all the package updates in there, though we generally recommend that you do). 

The distinction of "security updates" vs "normal updates" is mostly moot these days as people create CVEs for every single bug under the sun and there is no guarantee that a CVE fix will not cause a regression of some sort or change an expected behaviour.

If anything, the tendency of AL2 to install "security" updates on launch has caused more problems than it has improved anybody's security.

As for the specific issue with livepatches, I will make sure the kernel team has a look at it asap.

**Comment by puranjaymohan** [comment]
This was an issue with the dependency management between the kernel and the livepatch packages where when a kernel is removed, its corresponding livepatch package is not removed but downgraded to `1.0-0.amzn2023`. Now, when a new livepatch package version (example `1.0-1.amzn2023`) is released, `dnf upgrade` will try to install the new version of the livepatch package. This new version of the livepatch package has a dependency on the kernel, therefore `dnf` will try to install the kernel and remove another kernel version because `install_only` is set to 3. This cycle will keep triggering if there are orphan livepatch packages in the system.

This was fixed in `kernel-6.1.75-99.163.amzn2023`. So, the short term fix is to remove all livepatch packages that don't correspond to any of the 3 installed kernel versions before doing `dnf upgrade`

Once the system has all 3 installed kernel versions >= `kernel-6.1.75-99.163.amzn2023` then this issue will stop triggering.

**Comment by nosnilmot** [comment]
> Once the system has all 3 installed kernel versions >= `kernel-6.1.75-99.163.amzn2023` then this issue will stop triggering.

Thanks for confirming the root cause, and the manual corrective action required!
